Feature/config only

sebastianfeldmann opened the pull request:
This is a draft for the issue #1 

This way a Mediator implementation could look like this.

```php
<?php declare(strict_types=1);

namespace PharIo\Mediator;

class Plugin extends PluginBase
{
	/**
	 * Path to your mediator.json config file         
	 */
	protected $configPath  = '';

	/**
	 * If you are using a config file you have to remove the following lines otherwise settings
	 * defined here will overwrite settings defined in the configuration file.
	 */       
	protected $config = [
		// replace this with the your phar package name
		'packageName' => 'my-namespace/my-phar-package',
		// replace this with the path to your key directory
		'keyDirectory' => __DIR__ . '/../keys/',
		// replace this with the name of the binary that you want to use within the folder vendor/bin/
		'pharName' => 'my-binary',
	 	// replace this with the path to the phar file.
		// replacements are described in the [README.md](https://)
		'pharUrl' => 'https://github.com/heiglandreas/JUnitDiff/releases/download/%version%/my-binary.phar',
		// replace this with the path to the signature-file for the phar
		// replacements are described in the [README.md](https://)
		'pharSignatureUrl' => 'https://github.com/heiglandreas/JUnitDiff/releases/download/%version%/junitdiff.phar.asc'
	];
}
```

sebastianfeldmann commented:
Or if you have a `mediator.json` configuration file it would look something like this.

```php
<?php declare(strict_types=1);

namespace MyPharTool\ComposerPlugin;

class Plugin extends PluginBase
{
	protected $configPath  = __DIR__ . '/../mediator.json';
}
```

mediator.json
```javascript
{
	"packageName": "MyNameSpace/MyPharTool",
	"keyDirectory": "keys",
	"pharName": "my-phar",
	"pharUrl": "https://github.com/...",
	"pharSignatureUrl": "https://github.com/..."
}
```

heiglandreas commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
+.idea
```
I'd rather not have systemspecific stuff in the project-specific gitignore file. That is something for a users global gitignore file

sebastianfeldmann replied:
I removed the commit

heiglandreas commented on src/PluginBase.php:
```diff
@@ -24,6 +24,18 @@ abstract class PluginBase implements PluginInterface, EventSubscriberInterface
+	private static $defaultConfig = [
+		'keyDirectory'     => '',
+		'package'          => '',
+		'pharName'         => '',
```
We would need multiple entries for pharName,Url and signature for cases where multiple phars shall be installed (think phpcs and phpcbf)

sebastianfeldmann replied:
So `mediator.json` would look like this
```javascript
{
	"packageName": "foo/bar",
	"keyDirectory": "keys",
	"phars": [
		{
			"name": "foo",
			"file": "...",
			"signature": "..."
		}
	]
}
```

theseer replied:
Makes sense. And maybe it's overkill, but how about creating proper objects to encapsulate the values rather than having an ugly nested array? As in, after loading the JSON data.

sebastianfeldmann replied:
Yes after agreeing how the configuration should look we should definitely do that :)

heiglandreas commented on src/PluginBase.php:
```diff
@@ -64,5 +103,39 @@ public function createInstaller(string $pluginName, string $keyDirectory, Packag
+		$config     = [];
+		$fileValues = is_file($this->configPath)
+		            ? $this->loadConfig()
+		            : self::$defaultConfig;
```
Is it a clever idea to expect a variable here? What happens if the variable is not set?
Wouldn't an `abstract protected function getDefaultConfig(): Configuration` a better idea to make sure that a path is set?

heiglandreas reviewed (commented):
Great idea. But I was actually more thinking of leaving the current Abstract file as it is and put a separate class beneath it that will handle the config reading and stuff...

sebastianfeldmann commented:
I think I would actually go that far that we force a `mediator.json` configuration and skip the code configuration stuff completely.

I think you meat something like 

```php
class ConfiguredMediator extends PluginBase
{
    public abstract function getMediatorConfig(): string;

```

and then
```php
class MyPlugin extends ConfiguredMediator 
{
    public function getMediatorConfig(): string
    {
        return __DIR__ . '/../mediator.json';
    }
}
```

theseer commented on src/PluginBase.php:
```diff
@@ -24,6 +24,18 @@ abstract class PluginBase implements PluginInterface, EventSubscriberInterface
 
 	protected $io;
 
+	protected $configPath;
```
Is there a reason why this property is `protected` and not `private`?

sebastianfeldmann replied:
Was just a hacky way to setup the path :/

theseer commented on src/PluginBase.php:
```diff
@@ -24,6 +24,18 @@ abstract class PluginBase implements PluginInterface, EventSubscriberInterface
 
+	protected $configPath;
+
+	protected $config;
```
And this? ;)

Also: can we specify a type?

sebastianfeldmann replied:
Sadly not if we want to support 7.2 :(

theseer replied:
**mumbles**

`/** @var <type> */` then? pretty please? ;-)

heiglandreas commented:
I was more thinking of completely removing all code from the plugin if we go config only. The Plugin.php from the base project (composer-distributor) then retrieves the plugin name and folder from composer and checks that plugins folder for the mediator.json. Hopefully no need for a Plugin.php in the actual plugin folder...

theseer commented:
> I was more thinking of completely removing all code from the plugin if we go config only. The Plugin.php then retrieves the plugin name and folder from composer and checks that plugins folder for the mediator.json. Hopefully no need for a Plugin.php...

Wouldn't that be in the mediator? There has to be code somewhere and I believed that would be the distributor?

sebastianfeldmann commented:
> I was more thinking of completely removing all code from the plugin if we go config only. The Plugin.php from the base project (composer-distributor) then retrieves the plugin name and folder from composer and checks that plugins folder for the mediator.json. Hopefully no need for a Plugin.php in the actual plugin folder...

That would mean "the user" would have to register the vendor/ComposerDistributor class as a plugin.
So far I haven't found a way to retrieve the path of the installed package to look for the `mediator.json` without any code on the plugin side, @heiglandreas do you know of something?

sebastianfeldmann committed:
Allow config only setup

sebastianfeldmann commented:
I have not tried to further eliminate the whole plugin code I will investigate that further but for now this update has incorporated most of your feedback so far :)

- config value objects
- leave abstract class untouched
- remove hacky code configuration completely
- add type hints ;)

TODO: clean json config file handling

Thats how a plugin would look like.

```php
<?php

declare(strict_types=1);

namespace MyNamespace\MyTool;

use PharIo\ComposerDistributor\ConfiguredMediator;

class MyPlugin extends ConfiguredMediator 
{
    public function getMediatorConfig(): string
    {
        return __DIR__ . '/../mediator.json';
    }
}
```

mediator.json

```javascript
{
  "packageName": "foo/bar",
  "keyDirectory": "keys",
  "phars": [
    {
      "name": "foo",
      "file": "...",
      "signature": "..."
    }
  ]
}
```

theseer commented:
Looks good :) - except: Is there a reason for the method `getMediatorConfig()` to be public? Protected should be good enough, shouldn't it? :)

sebastianfeldmann commented:
I added the missing json config validation and changed the json error handling because PHP7.2 :)

theseer commented:
Looks pretty awesome so far! :+1: 

Two questions popped up into my mind, will add them as review comments...

theseer commented on src/ConfiguredMediator.php:
```diff
@@ -0,0 +1,104 @@
+		foreach ($config->phars() as $phar) {
+			$files[] = new File(
+				$phar->name(),
+				Url::fromString($phar->file()),
+				Url::fromString($phar->signature())
```
Are these always URLs? If so, we could move these calls into the Config and only return URLs rather than plain strings. That would avoid the need to handle exceptions from `Url::fromString` in the case the string value is b0rked...

sebastianfeldmann replied:
I can't think of any case where something other than an URL would make sense, I'll move it to the configuration.

sebastianfeldmann replied:
Done

theseer commented:
Question two has to go here anyhow ;)

We now have very good looking parsing code to deal with the json file. I'm not convinced though it belongs to the class it currently is in but should be extracted into its own class.

Thoughts?

sebastianfeldmann commented:
> Question two has to go here anyhow ;)
> 
> We now have very good looking parsing code to deal with the json file. I'm not convinced though it belongs to the class it currently is in but should be extracted into its own class.
> 
> Thoughts?

Thanks the question itself helps a lot, I thought so myself but wan't sure either. I will extract the json load and validation stuff into a `Config\Reader`

heiglandreas commented:
> > I was more thinking of completely removing all code from the plugin if we go config only. The Plugin.php from the base project (composer-distributor) then retrieves the plugin name and folder from composer and checks that plugins folder for the mediator.json. Hopefully no need for a Plugin.php in the actual plugin folder...
> 
> That would mean "the user" would have to register the vendor/ComposerDistributor class as a plugin.
> So far I haven't found a way to retrieve the path of the installed package to look for the `mediator.json` without any code on the plugin side, @heiglandreas do you know of something?

Nope. But to be honest: I haven't looked so far ;-)

But so far the current setup is pretty "low-code" already. So from my side I'm happy with that as a starting point!

sebastianfeldmann commented:
I moved the json parsing stuff to `Config\Factory`.
Another thing I'm unsure about is if we should use `json_decode` to `stdClass` or  to `array` ?

I think `array` fits better because passing a `stdClass` to `Factory::createConfig` feels weird.
`array` would have the charm of in place configuration.

```php
$factory = new Config\Factory();
$config  = $factory->createConfig(["packageName": "phar-io/phive",......]);
```

What do you think?

heiglandreas commented on src/Config/Factory.php:
```diff
@@ -0,0 +1,85 @@
+	 * }
+	 *
+	 */
+	public function createFromFile(string $configFile): Config
```
Shouldn't this be `static`?

sebastianfeldmann replied:
UPS #fixed

heiglandreas commented on src/Config/Phar.php:
```diff
@@ -0,0 +1,47 @@
+	 */
+	private $signature;
+
+	public function __construct(string $name, string $file, string $signature)
```
Can't we expect `Url` objects here already?

sebastianfeldmann replied:
I completely replaced `Phar` with `File`

heiglandreas commented on src/ConfiguredMediator.php:
```diff
@@ -0,0 +1,45 @@
+
+		$files = [];
+		foreach ($config->phars() as $phar) {
+			$files[] = new File(
```
Why do we need to create a new Class here? Can't we use the `Phar` class here instead? Or use a `File::createFromPharObject()` method? Or on the other hand use the `File` instead for the configuration?

sebastianfeldmann replied:
Good Question, I will investigate :)

sebastianfeldmann replied:
I removed the `Phar` class and replaced it with `File`and changed the return type from `Config::phars` to `FileList`

heiglandreas commented:
> I moved the json parsing stuff to `Config\Factory`.
> Another thing I'm unsure about is if we should use `json_decode` to `stdClass` or to `array` ?
> 
> I think `array` fits better because passing a `stdClass` to `Factory::createConfig` feels weird.
> `array` would have the charm of in place configuration.
> 
> ```
> $factory = new Config\Factory();
> $config  = $factory->createConfig(["packageName": "phar-io/phive",......]);
> ```
> 
> What do you think?

I stumbled over that as well and it felt weird. TypeHinting stdClass or array is both kind of not saying anything, so from that point it is irrelevant. But the "in place config" would verify using array instead of stdClass.

theseer commented:
Given that keys in json can contain chars not allowed of PHP classes or properties, the Array format is what I usually prefer. If you have bizzare key names, the PHP way to access them anyhow is `$json->{"foo something"}` which is really ugly ;)

Let's just settle with array.

sebastianfeldmann commented:
Switched `json_decode`  to ASSOC and setup is handled with arrays now

sebastianfeldmann commented:
I updated comments, file headers and type hints to a format I found in `phive` .

theseer commented on src/Config/Factory.php:
```diff
@@ -0,0 +1,108 @@
+		if (!is_file($configFile)) {
+			throw new \RuntimeException('Config file is missing');
+		}
+		$json = json_decode(file_get_contents($configFile), true);
+		if (json_last_error() !== JSON_ERROR_NONE) {
+			throw new \RuntimeException('Invalid json configuration');
+		}
```
If we drop PHP 7.2 support - which will be EOL in 17 days - and instead require PHP 7.3+, we could use `json_decode(....., JSON_THROW_ON_ERROR);` rather than throwing a generic `RuntimeException`..

sebastianfeldmann replied:
Done

theseer commented on src/Config/Factory.php:
```diff
@@ -0,0 +1,108 @@
+	 * }
+	 *
+	 */
+	public static function createFromFile(string $configFile): Config
```
While I do understand the name of the method, technically it implies that the factory is createdFromString ;-)

Maybe we rename the thing into `Loader` and than can use `loadFile()`?

sebastianfeldmann replied:
I separated json loading and config creation in two separate classes `Config\Loader` and `Config\Factory` 
What do you think?

theseer replied:
Technically, the `Factory` is a `Mapper` then, but yes ;)

sebastianfeldmann replied:
So we keep it :)

theseer replied:
Well, a `Mapper` is good thing, not a bad thing.

Maybe it's a personal thing, but a `Factory` is something else for me. This thing merely translates an array into a DTO. That's exactly what mappers do ;)

sebastianfeldmann replied:
Rename it is ;)

sebastianfeldmann replied:
Done

theseer commented on src/Config/Factory.php:
```diff
@@ -0,0 +1,108 @@
+		return new Config($json['package'], $json['keyDir'], $this->createPhars($json));
+	}
+
+	private function createPhars(array $json): FileList
```
Why don't we just pass in $json['phars'] here?

sebastianfeldmann replied:
Fixed

theseer commented:
Side note observation: While it's a lot of fun doing this, how can we possibly spent so much time tweaking such a small amount of code? ;-)

sebastianfeldmann commented:
> Side note observation: While it's a lot of fun doing this, how can we possibly spent so much time tweaking such a small amount of code? ;-)

Cause we want to do it right? :) 
But I think it is fine now (hopefully)

heiglandreas commented on src/Config/Config.php:
```diff
@@ -0,0 +1,41 @@
+	/** @var string */
+	private $package;
+
+	/** @var string */
```
Can't we use an `SPLFile`-object here? Thast way we can more easily check whether the file or folder actually exists and raise an Exception

heiglandreas commented on src/Config/Config.php:
```diff
@@ -0,0 +1,41 @@
+
+	public function __construct(string $package, string $keyDir, FileList $phars)
+	{
+		$this->package      = $package;
```
shouldn't we do a basic sanity check here? Like all lower caps and contains one `/`?

sebastianfeldmann replied:
Done

heiglandreas commented on src/Config/Config.php:
```diff
@@ -0,0 +1,41 @@
+	public function __construct(string $package, string $keyDir, FileList $phars)
+	{
+		$this->package      = $package;
+		$this->keyDirectory = $keyDir;
```
And here perhaps at least a check whether the folder or file actually exists?

sebastianfeldmann replied:
Done

heiglandreas commented on src/Config/Mapper.php:
```diff
@@ -0,0 +1,79 @@
+		if (!isset($pharData['file']) || !is_string($pharData['file'])) {
+			throw new \RuntimeException('Invalid phar config  \'file\' is missing');
+		}
+		if (!isset($pharData['signature']) || !is_string($pharData['signature'])) {
```
I would treat the signature as optional. If someone wants to distribute their file as a binary but has trouble signing it (for whatever reason) they will not have a signature. We should still allow them to use this feature. And there will be a note on installation that the tool was installed without a signature and that people will use it entirely at their own risk!

sebastianfeldmann replied:
In that case `keyDirectory` becomes optional as well right?

heiglandreas replied:
Absolutely!

sebastianfeldmann replied:
So we should change the return type to `?SplFile`

sebastianfeldmann replied:
Done

heiglandreas commented:
If we go mediator.json only we might be able to modify the installer instantiation so that it can use the configuration objects instead of strings. something like 

```php
public function installOrUpdateFunction(PackageEvent $event): void
{
	$config    = Config\Loader::loadFile($this->getMediatorConfig());
	$installer = self::fromConfig(Config $config);
	$installer->install();
}
```

sebastianfeldmann committed:
Add missing parameter

sebastianfeldmann commented:
- `Config` return value changed from `string` to `SplFileInfo`
- Added `ConfiguredMediator::createInstallerFromConfig`
- `File` changed so `signature` is now optional.
- `Installer` changed so `KeyDirectory` is now optional

sebastianfeldmann committed:
Allow install without verification

For this to be possible config values for 'keyDirectory'
and 'signature' have to be optional.

heiglandreas commented on src/Config/Config.php:
```diff
@@ -0,0 +1,44 @@
+	/** @var \PharIo\ComposerDistributor\FileList */
+	private $phars;
+
+	public function __construct(string $package, ?\SplFileInfo $keyDir, FileList $phars)
```
Sollten wir die Reihenfolge der Parameter nicht umdrehen dann?

sebastianfeldmann replied:
And default `$keyDir` to null?

sebastianfeldmann replied:
Done

heiglandreas committed:
Fix wrong parameter name

The parameter-name was typed wrongly and needed to be fixed here

heiglandreas committed:
Fix Package-Retrieval for Composer 2.0

In composer 2.0 the way to retrieve the current packager changed so this fixes
the issue by checking whether we are using composer v1 or composer v2 and using
the appropriate way to retrieve the current package.

heiglandreas committed:
Add integration-tests

This adds 3 tests that check whether the actual integration works.

Currently the test checking whether a signature is valid with a wrong key
doesn't fail. I'll investigate that!

heiglandreas committed:
Fix tests

This removes some more temporary files and checks the correct signature

heiglandreas committed:
Distinguish between integration and unit-tests

To make it easier to distinguish between the unit-tests and the integration
tests.

sebastianfeldmann committed:
Merge pull request #7 from phar-io/feature/config-only2

Distinguish between integration and unit-tests

sebastianfeldmann committed:
Add tools and build files
